[dynamo] Forward OptimizedModule.__setattr__ to the wrapped module

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122098

Fixes #114844

In the linked issue we have
```
compiled_module = torch.compile(module)
compiled_module.x = ...
compiled_module(...)  # Mutates self.x
```
Where since the module mutates `self.x` you would expect `compiled_module.x`
to be updated but actually `compiled_module.x = ...` sets an attribute "x"
on the `OptimizedModule` object while the forward method of the module mutates
`module.x`.

This gives the expected behavior by forwarding `compiled_module.__setattr__`
down to `module.__setattr__`. There is already a corresponding `__getattr__`
so now `compiled_module.x` becomes an alias for `module.x`.

cc @ezyang @gchanan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 commented on torch/_dynamo/debug_utils.py:
```diff
@@ -321,19 +319,9 @@ def run_fwd_maybe_bwd(gm, args, only_fwd=False, disable_clone=False):
-    if not hasattr(gm, "_boxed_call") and (
-        orig_named_parameters is not None or orig_named_buffers is not None
-    ):
-        gm = make_boxed_func(gm)
```
I removed the `make_boxed_func` wrapper here as it was interfering with the `collect_results` call below which does `isinstance(gm, OptimizedModule)`.

peterbell10 commented on torch/_dynamo/debug_utils.py:
```diff
@@ -359,21 +347,8 @@ def same_two_models(
-
-    if isinstance(opt_gm, OptimizedModule):
-        opt_gm.named_parameters = named_parameters_for_optimized_module(opt_gm)
-        opt_gm.named_buffers = named_buffers_for_optimized_module(opt_gm)
```
This is aliasing `gm.named_parameters` and `gm.named_buffers` to the underlying `gm._orig_mod` attributes,  which is no longer necessary.

peterbell10 committed:
Update

[ghstack-poisoned]

ezyang reviewed (approved):
very nice

lezcano commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but the distributed failure looks legit as it is also failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/88ebdbc97c103271766203df6662240e95a09b42' -c ignoredsignal 

```
2024-03-18T19:44:20.8615564Z FAILED [6.3146s] distributed/checkpoint/test_state_dict.py::TestStateDict::test_fsdp2 - RuntimeError: Process 0 exited with error code 10 and exception:
2024-03-18T19:44:20.8615744Z Traceback (most recent call last):
2024-03-18T19:44:20.8616656Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 656, in run_test
2024-03-18T19:44:20.8616811Z     getattr(self, test_name)()
2024-03-18T19:44:20.8617718Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 542, in wrapper
2024-03-18T19:44:20.8617843Z     fn()
2024-03-18T19:44:20.8618693Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2739, in wrapper
2024-03-18T19:44:20.8618858Z     method(*args, **kwargs)
2024-03-18T19:44:20.8619863Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 374, in wrapper
2024-03-18T19:44:20.8620109Z     func(self, *args, **kwargs)  # type: ignore[misc]
2024-03-18T19:44:20.8620995Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 181, in wrapper
2024-03-18T19:44:20.8621154Z     return func(*args, **kwargs)
2024-03-18T19:44:20.8621843Z   File "/var/lib/jenkins/workspace/test/distributed/checkpoint/test_state_dict.py", line 264, in test_fsdp2
2024-03-18T19:44:20.8621985Z     self.run_subtests(
2024-03-18T19:44:20.8623138Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 353, in run_subtests
2024-03-18T19:44:20.8623374Z     return run_subtests(self, *args, **kwargs)
2024-03-18T19:44:20.8624424Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 408, in run_subtests
2024-03-18T19:44:20.8624681Z     test_fn(*test_args, **test_kwargs, **subtest_kwargs)
2024-03-18T19:44:20.8625374Z   File "/var/lib/jenkins/workspace/test/distributed/checkpoint/test_state_dict.py", line 259, in _test_fsdp2
2024-03-18T19:44:20.8625557Z     self._test_save_load(init_model_optim)
2024-03-18T19:44:20.8626275Z   File "/var/lib/jenkins/workspace/test/distributed/checkpoint/test_state_dict.py", line 140, in _test_save_load
2024-03-18T19:44:20.8626447Z     dist_msd = dist_model.state_dict()
2024-03-18T19:44:20.8627252Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1911, in state_dict
2024-03-18T19:44:20.8627854Z     module.state_dict(destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
2024-03-18T19:44:20.8628543Z TypeError: _patch_model_state_dict.<locals>.state_dict_call() got an unexpected keyword argument 'destination'
```

peterbell10 committed:
Update on "[dynamo] Forward OptimizedModule.__setattr__ to the wrapped module"


Fixes #114844

In the linked issue we have
```
compiled_module = torch.compile(module)
compiled_module.x = ...
compiled_module(...)  # Mutates self.x
```
Where since the module mutates `self.x` you would expect `compiled_module.x`
to be updated but actually `compiled_module.x = ...` sets an attribute "x"
on the `OptimizedModule` object while the forward method of the module mutates
`module.x`.

This gives the expected behavior by forwarding `compiled_module.__setattr__`
down to `module.__setattr__`. There is already a corresponding `__getattr__`
so now `compiled_module.x` becomes an alias for `module.x`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

atalman commented:
@pytorchbot revert -m "Failing internally" -c ghfirst

peterbell10 commented:
@atalman can you share any details about the failure?

ezyang commented:
The test failure looks like this

```
ERROR: test_dynamo_walk_module (caffe2.torch.fb.module_factory.sync_sgd.tests.test_torchdynamo_wrapper.TorchDynamoWrapperTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/module_factory/sync_sgd/tests/test_torchdynamo_wrapper.py", line 163, in test_dynamo_walk_module
    self.assertEqual({x for x, y in params}, expected_params)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/module_factory/sync_sgd/tests/test_torchdynamo_wrapper.py", line 163, in <setcomp>
    self.assertEqual({x for x, y in params}, expected_params)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 377, in module_walk_parameters
    yield from cast(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 234, in module_walk
    for module_name, module in _walk_proxy_modules(model):
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 88, in _walk_proxy_modules
    yield from _walk_proxy_modules(child, f"{prefix}{child_name}")
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 85, in _walk_proxy_modules
    yield from _walk_proxy_modules(child, new_prefix)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 78, in _walk_proxy_modules
    _, child = get_descendant_module_from_str(module, child_path)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 167, in get_descendant_module_from_str
    module = module._modules[child_name]
KeyError: '_orig_mod'

```

ezyang commented:
The code performing the walking is this:

```
def _walk_proxy_modules(
    module: nn.Module,
    prefix: str = "",
    custom_named_children_fn: Optional[str] = CUSTOM_NAMED_CHILDREN_FN,
) -> Iterator[Tuple[str, nn.Module]]:
    r"""
    This traversal function walks through all the submodules of the given model
    and returns:
    1.) all the modules that have no proxy module as its ancestor,
    2.) all the proxy modules in the model.

    It can also take a custom_named_children_fn to override module.named_children, in
    order to customize the behavior of accessing immediate children.
    custom_named_children_fn on module is expected to have the signature () -> Dict[str, str],
    where the key is a path to descendant, and value is the new name given to it.
    """
    yield prefix, module
    if not ProxyModule.is_proxy_module(module) and not isinstance(
        module, TorchRecShardedModule
    ):
        # No need to walk into proxy module or torchrec's sharded module
        prefix = f"{prefix}." if prefix else ""

        # Some modules like FSDP overloads getattr() and gets attr from the children.
        # Check with dir() to make sure it indeed has the attr.
        if (
            custom_named_children_fn is not None
            and hasattr(module, custom_named_children_fn)
            and custom_named_children_fn in dir(module)
        ):
            named_children_fn = getattr(module, custom_named_children_fn)
            for child_path, child_name in named_children_fn().items():
                _, child = get_descendant_module_from_str(module, child_path)
                if not child_name:
                    # if child_name is empty, it means we want to skip its name,
                    # so we use the original prefix passed in and remove the trailing dot
                    new_prefix = prefix[:-1]
                else:
                    new_prefix = f"{prefix}{child_name}"
                yield from _walk_proxy_modules(child, new_prefix)
        else:
            for child_name, child in module.named_children():
                yield from _walk_proxy_modules(child, f"{prefix}{child_name}")
```

ezyang commented:
It is likely that `named_children_fn` is this `optimized_child.named_children_walk = lambda: {"_orig_mod": ""}`

So now the problem makes sense: the internal test failure appears to be assuming that an OptimizedModule has an `_orig_mod` attribute which is a submodule. But now orig_mod is no longer being registered to the _modules dictionary for some reason (maybe because setattr is supposed to do it, and your code changed this), so now it fails.

I hope this is enough to debug the rest.

peterbell10 commented:
> the internal test failure appears to be assuming that an `OptimizedModule` has an `_orig_mod` attribute which is a submodule. But now `orig_mod` is no longer being registered to the `_modules` dictionary for some reason 

This doesn't seem to make sense with what I see locally, as the compiled module does have an `_orig_mod` attribute and it's listed in `_modules`:
```python
import torch

mod = torch.nn.BatchNorm2d(10, device="cuda")
a = torch.randn([1, 10, 5, 5], device="cuda")
cmod = torch.compile(mod)

print(f"{=:type(cmod)}")           # <class 'torch._dynamo.eval_frame.OptimizedModule'>
print(list(cmod.named_children())) # [('_orig_mod', BatchNorm2d(...))]
print(cmod._modules)               # OrderedDict([('_orig_mod', BatchNorm2d(...))])
print(mod._modules)                # OrderedDict()
```

which is the same result as before this PR.

> It is likely that `named_children_fn` is this `optimized_child.named_children_walk = lambda: {"_orig_mod": ""}`

If this is true, then the error does make sense as this would set the `named_children_walk` function on `mod` rather than `cmod`. So the walk would expect there to be a `mod._orig_mod`.

I don't really see how we could work around this, so long as `cmod` is pretending to be `mod` then monkey-patching like this isn't going to work quite as expected. Though the monkey-patching could be done by setting `cmod.__dict__` directly.

I'm not sure in context though, is this custom walk function some kind of workaround for an issue with walking optimized modules which is fixed in this PR or is there some other reason not to use `cmod.named_children()`?

ezyang commented:
I have no idea what the code here does or why it is written this way haha. If you think that we cannot first principles fix the BC break, let's patch the code. Maybe if we add `named_children_walk` to the not pass through list?

ezyang commented:
It fixes the test.

```
diff --git a/fbcode/caffe2/torch/_dynamo/eval_frame.py b/fbcode/caffe2/torch/_dynamo/eval_frame.py
--- a/fbcode/caffe2/torch/_dynamo/eval_frame.py
+++ b/fbcode/caffe2/torch/_dynamo/eval_frame.py
@@ -125,6 +125,7 @@
         "forward",
         "_forward",
         "__dict__",
+        "named_children_walk",
     }
 
     def __init__(self, mod: torch.nn.Module, dynamo_ctx):
```

ezyang committed:
Update

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge -r

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
